### PR TITLE
Change -maxtimeadjustment default from 70 minutes to 0

### PR DESCRIPTION
**Pull request by jonatack**
Following the core dev IRC discussion today, this is a straightforward step per the suggestion at https://www.erisian.com.au/bitcoin-core-dev/log-2022-03-17.html#l-652.  Users can still opt in to adjusted time by setting `-maxtimeadjustment` with a value greater than zero. This allows gaining feedback as to what extent user space depends on adjusted time.

If we want to remove the adjusted time code down the road, it makes sense to give users the fallback option to use adjusted time for a release or two. This also reduces our code reversion risk in the case that users rely on it. The release notes would describe the change and suggest that the option is being deprecated for future removal.

See #7573 and #4521 for further context.

### CODE REVIEW

**Comment by jonatack** [comment]
Will add a release note if concept/approach here is good.

**Comment by jonatack** [comment]
Ah, it looks like https://github.com/bitcoinxt/bitcoinxt/pull/140 did the same.

The default remains 0 since: https://github.com/bitcoinxt/bitcoinxt/blob/48bbdd48e71af98c467caf31cbf93e8339e2ae7e/src/timedata.h#L13

**Comment by jonatack** [comment]
Dropped the doc commit (proposed separately in #24609 to potentially be added to #24512) and moving this out of draft.

**Review by vasild — COMMENTED** [review]
Concept ACK

The time data comes from untrusted peers over cleartext, unauthenticated and unencrypted, so it could be bogus. However, I guess, the assumption here is that most of the nodes are honest. Note that we choose those 200 peers randomly from our addrman, sort their datas and only take the median. So, an evil actor who wants to mess with our perceived time has to control >100 of those peers, or >50% of our addrman entries.

What is the most useful thing that could happen if the system clock is wrong indeed during `bitcoind` startup and it gets adjusted by those 200 honest samples (in a good way)? Or in other words - what will go wrong in that case if we refuse to adjust the clock?

**Comment by MarcoFalke** [comment]
I am not sure this is the right way to go. If we think the changes are ok to make, we should just remove adjusted time. If we don't, then we shouldn't do the changes here either.

No objection to the changes here, but to me it seems to combine the worst of everything:

* It disables the clumsy code by default, making it effectively dead code, since almost no user changes the default values.
* It doesn't give any benefits to modules that use adjusted time, as they still need to assume adjusted time may be wrong, even though the system time is correct.
* User that realize they might need this setting (and change the default) might easier just set the correct system time. Or are you going to write release notes saying: "If your system clock is wrong, please make sure to set -maxtimeadjustment to the value your clock is wrong at most."  Instead of: "If your clock is wrong, correct it".

**Comment by luke-jr** [comment]
Concept ACK. Seems like a good one-release-only stage before complete removal?

**Comment by jonatack** [comment]
@luke-jr yes, this seems like the right thing to do for users.

**Comment by jonatack** [comment]
@MarcoFalke yes, this change doesn't preclude ripping out the code permanently afterward. if I understand correctly, this code has been in the codebase since 2016, roughly six years now. If there's no urgent need to rip it all out, which may be cleaner for us but is a larger change, it makes sense to give users the fallback option to use adjusted time for a release or two. I don't feel competent dot second-guess what users are doing or not doing. The release notes would describe the change and suggest that the option is being deprecated for future removal.

**Comment by jonatack** [comment]
@MarcoFalke indeed, this change doesn't preclude ripping out the code permanently afterward. if I understand correctly, this code has been in the codebase since 2016, roughly six years now. If there's no urgent need to rip it all out, which may be convenient for us but is a larger change, it may make sense to give users the fallback option to use adjusted time for a release or two. I don't feel competent to second-guess what users are doing or not doing. The release notes would describe the change and suggest that the option is being deprecated for future removal.

File: src/test/timedata_tests.cpp [hunk 1]
```diff
@@ -16,7 +16,16 @@
...
 #include <boost/test/unit_test.hpp>
 
-BOOST_FIXTURE_TEST_SUITE(timedata_tests, BasicTestingSetup)
+/**
+ * Set -maxtimeadjustment to 70 minutes for testing purposes; this was
+ * previously the default value before being changed to 0 in v24.0.
+ */
+struct TimeDataTestingSetup : public BasicTestingSetup {
+    TimeDataTestingSetup()
+        : BasicTestingSetup{CBaseChainParams::MAIN, /*extra_args=*/{"-maxtimeadjustment=4200"}} {}
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/test/timedata_tests.cpp:25]: Thanks, done.

**Comment by MarcoFalke** [comment]
If there is any user using this with a valid use case in mind, we shouldn't remove it nor deprecate it. In that case, we should probably fix it so that users with a proper system clock aren't negatively affected by adjusted time and the users with an improper system clock are covered in all/most cases where it is needed. So it would be good to know if any users are using this, and I think the most effective way to find out is by removing it (User will see an init error on startup, if they are using it). I am not sure how many users would notice with just the default value changed and a release note snippet. An alternative would be to add a `-settingdeprecated=` setting, similar to `-rpcdeprecated=`?

Again, I am not against the changes here. They look fine and I don't see any harm in them. I am only wondering how much benefit they will bring.

**Comment by michaelfolkson** [comment]
> If there is any user using this with a valid use case in mind, we shouldn't remove it nor deprecate it.

If this is indeed the approach of the project today nothing should ever be deprecated (including say the legacy wallet in #20160) because you can never be sure a user isn't using it. I'm not sure when this switch to never deprecating anything happened. There are various examples (e.g. #8780 and more recent examples) in the past of things being deprecated over multiple versions. Of course giving time to developers/users to prepare for the changes and good communication are important.

Maybe something to bring up in the Core dev meeting this week.

**Comment by jonatack** [comment]
@michaelfolkson let's please stay on this topic. RPC result fields (and sometimes input params) are still deprecated and a config option may possibly be removed (having a deprecation process for this may be an improvement). Removing an RPC outright seems much more rare to avoid needlessly breaking user space, and sometimes an RPC is made hidden but not removed.

**Comment by vasild** [comment]
@mzumsande, good points! I think 1. and 2. can be done in parallel. If 1. is fully completed before 24.x then 2. becomes irrelevant. Having 2. now could help with some testing by people that use pre-release, and don't change the defaults, I guess.

**Comment by laanwj** [comment]
Concept ACK.

**Comment by jonatack** [comment]
> So it would be good to know if any users are using this, and I think the most effective way to find out is by removing it.

The least risky way to find out is to do this for a release or so, because it is trivial to revert.

>  is there much of a difference between 1) and 2) if it all happens for the 24.0 release (which seems realistic?)

The risk and ease of reversion.

**Comment by MarcoFalke** [comment]
> The least risky way to find out is to do this pull for a release or so, because it is trivial to revert and gives users optionality.

If users change the default setting, they won't find out if this pull is merged. But maybe I am just overthinking this, as almost no one changes the default values.

**Comment by jonatack** [comment]
True, in any case we need to communicate about it.

Devil's advocate, if there is a potential attack vector with the current code that should be closed, that would be a different scenario.

File: src/timedata.h [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #include <stdint.h>
 #include <vector>
 
-static const int64_t DEFAULT_MAX_TIME_ADJUSTMENT = 70 * 60;
+static constexpr int64_t DEFAULT_MAX_TIME_ADJUSTMENT{0};
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/timedata.h:13]: Good eye; added a commit authored by you.

**Comment by fanquake** [comment]
I think I agree with @MarcoFalke & @mzumsande and would rather move ahead with refactoring the code, i.e #24662 + validation changes, rather than a sort of interim / soft deprecation period of just changing the default. I think we have a long enough window prior to the 24.x branch off to get the refactoring done, and if for some reason that didn't happen, we could always, if still wanted, make this change just prior to branch off.

I would also be very surprised if anyone is actually using this setting in production.

**Comment by jonatack** [comment]
> I think I agree with @MarcoFalke & @mzumsande and would rather move ahead with refactoring the code, i.e #24662 + validation changes, rather than a sort of interim / soft deprecation period of just changing the default. I think we have a long enough window prior to the 24.x branch off to get the refactoring done, and if for some reason that didn't happen, we could always, if still wanted, make this change just prior to branch off.
> 
> I would also be very surprised if anyone is actually using this setting in production.

A headstrong and presumptive approach would needlessly take on project and user space risk. Thus the straightforward proposal here, unless there is some sort of new urgency.

File: src/test/timedata_tests.cpp [hunk 3]
```diff
@@ -16,7 +16,16 @@
 
 #include <boost/test/unit_test.hpp>
 
-BOOST_FIXTURE_TEST_SUITE(timedata_tests, BasicTestingSetup)
+/**
+ * Set -maxtimeadjustment to 70 minutes for testing purposes; this was
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/test/timedata_tests.cpp:20]: For a few of the tests it might make sense, if this approach goes forward.

**Comment by mruddy** [comment]
I just went through the old conversation for #7573 where this flag was added and I found the reason why we did not set the default to zero back [then:](https://github.com/bitcoin/bitcoin/pull/7573#issuecomment-188844279) "I chose not to disable the time-offset logic entirely because I've gotten feedback about people liking to get the warning message "Please check that your computer's date and time are correct!" if none of the peers are within 5 minutes just in-case."
And: "Right now, if the median offset is more than 70 minutes, the offset is set to zero and a warning is logged if none of the nodes are within five minutes. So, there is some protection against an attack that attempts to change your node's perspective on time a lot. If your node disagrees substantially with every other node, then you get a warning. But you'll continue to exchange transactions and blocks etc..."
This was the justification for setting the flag to zero, when it is used: https://github.com/bitcoin/bitcoin/pull/7573#issuecomment-192626502
But, this comment also supports not setting the default to zero: https://github.com/bitcoin/bitcoin/pull/7573#issuecomment-192584061

In summary, I tend to Concept NACK this changing of default for the reasons mentioned.

**Comment by jonatack** [comment]
Hi @mruddy, this change is proposed within the context of a process to remove adjusted time completely.

**Comment by mruddy** [comment]
When I think about removing adjusted time completely, this quote from sipa comes to mind:
"I am not convinced that the behaviour of not correcting at all is less harmful than correcting your clock based on what random peers tell you when chosen by someone who has confidence in their own clock configuration."
https://github.com/bitcoin/bitcoin/pull/7573#issuecomment-192584061

I think there is some subtle wisdom in sipa's sentiment at the time and I think it still applies. Initially, when I proposed the PR that led to the creation of the `-maxtimeadjustment` flag, my thought was something similar to, "I just want to be able to ignore what other nodes are telling me, because I know better, and I've configured my system." Atfer considering what sipa and gmaxwell wrote in that PR's comments, I changed my mind to be more accepting that adjusted time could still be useful for a lot of less actively/professionally managed nodes. I think that sipa's and gmaxwell's wisdom were based on experiences of NTP failing silently, NTP being insecure and spoofable, lack of node upkeep/monitoring, and probably more.

I still think that adjusted time can serve the same purpose as it did in 2016. I still think that the justification for the `-maxtimeadjustment` flag is still valid too (https://github.com/bitcoin/bitcoin/pull/7573#issuecomment-192626502). So, I still lean to the Concept NACK on this. But, I'm open to change my mind if there's a new persuasive reason to nuke adjusted time.

**Comment by mruddy** [comment]
I think that if we want to progress with removing the mechanism that adjusts our node's view of time, that we should still keep a mechanism that warns about being out of alignment with peer times.
Thus, we'd warn, but not correct (similar to how we warn when we do not adjust, and no other peer is within 5 minutes of our clock).
We'd want to avoid being too verbose when logging this so that we avoid hostile p2p version messages that are designed just to spam our logs.

That would allow us to keep clock sync as a local system problem.
When the system clock is off, one might run into a bunch of different problems. And, it's a bigger issue than just syncing or checking NTP, for example https://tails.boum.org/contribute/design/Time_syncing/
Also, here's some data, more current than when #7573 was worked, on how far off clocks on internet hosts can be https://labs.apnic.net/?p=1186

So, if this PR is moving in that direction, then I'd be ok with concept ack'ing. But, getting to the actual change, only changing the default to 0, seems kind of lacking. It gets closer to the end goal, but it leaves a lot of clumsy code to get log messages. So, I tend to agree with this https://github.com/bitcoin/bitcoin/pull/24606#issuecomment-1072526795

**Comment by MarcoFalke** [comment]
Not sure how useful it is to look at "internet hosts" to get data points, which are off by several days. (I haven't checked, but) Bitcoin Core should refuse to start up if the clock is behind the latest block saved on disk (or will refuse to download it). It is certainly not uncommon for some clocks to be broken. I've run into this myself, see for example https://bugs.launchpad.net/cloud-images/+bug/1896443 . However, Bitcoin Core will already fail to function properly on those systems and adjusted time doesn't correct offsets in the order of days anyway.

I agree on the warnings. Someone should check that Bitcoin Core raises an InitError when the clock is behind the latest block saved on disk (or maybe the latest data point in the chainTxData). I proposed more aggressive warnings or errors in a recent IRC meeting, but there was no agreement that more aggressive warnings are warranted. So I think we should check that the existing warnings are properly sent out even if the default is set to 0 or the setting removed.

**Comment by MarcoFalke** [comment]
There is a meta issue if anyone feels like contributing data points: #24805

**Comment by MarcoFalke** [comment]
Another idea for improving the warning: We could calculate the running median offset of the last 10 outbound peer connections instead of the current offset, which locks in after (and takes into account) 200 outbound connections?

**Comment by jonatack** [comment]
This won't be merged, closing.
